src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/AdditionalCondition14bHelper.kt

```diff
@@ -0,0 +1,191 @@
+package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.service
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.entity.AdditionalCondition
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.entity.AdditionalConditionData
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.entity.Licence
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.LicenceStatus
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
+private val dateTimeFormatter = DateTimeFormatter.ofPattern("EEEE dd MMMM yyyy")!!
+private const val conditionCode14B = "524f2fd6-ad53-47dd-8edc-2161d3dd2ed4"
+private const val ledDateChangeMsg = "licence end date"
+private const val crdOrArdChangeMsg = "release date"
+private const val ledAndCrdOrArdChangeMsg = "release date and licence end date"
+private const val END_Date = "endDate"
+
+fun updateAdditionalConditionWithAdditionConditionData(
+  sentenceChanges: SentenceChanges,
+  updatedLicenceEntity: Licence,
+  licenceEntity: Licence
+): AdditionalCondition? {
+  var endDate: String? = null
+  val additionalCondition14b = hasExistingLicenseContains14B(licenceEntity)
+  additionalCondition14b?.let {
+    if (hasCorrectStatusCodeForExisting14B(licenceEntity)) {
+      endDate = calculateEndDateForLedArdCrdChanges(sentenceChanges, updatedLicenceEntity, licenceEntity)
+    }
+    if (!endDate.isNullOrBlank()) {
+      val additionalConditionDataWith14bEndDate = additionalCondition14b.additionalConditionData
+        .find { it.dataField == END_Date }?.copy(
+          dataValue = endDate
+        )
+      return additionalCondition14b.copy(
+        conditionVersion = updatedLicenceEntity.version!!,
+        additionalConditionData = additionalCondition14b.additionalConditionData.removeExistingEndDateAndAddNew(
+          additionalConditionDataWith14bEndDate!!
+        ),
+        expandedConditionText = additionalCondition14b.conditionText?.replace("[INSERT END DATE]", endDate!!)
+      )
+    }
+  }
+  return null
+}
+fun addAdditionalConditionDatasToAdditionalCondition(
+  conditionCode: String,
+  licenceEntity: Licence,
+  newAdditionalCondition: AdditionalCondition
+): List<AdditionalConditionData>? {
+  return when {
+    isAdditionalConditionOf14BWithCorrectStatusCode(conditionCode, licenceEntity) -> createAdditionalConditionDataFor14b(licenceEntity, newAdditionalCondition)
+    else -> null
+  }
+}
+fun getReasonForDateChange(
+  hasLedChange: Boolean,
+  updatedLicenceEntity: Licence,
+  licenceEntity: Licence
+): String? {
+  val hasArdOrCrdChanged = hasArdOrCrdChanged(updatedLicenceEntity, licenceEntity)
+  return when {
+    hasLedChange && hasArdOrCrdChanged -> ledAndCrdOrArdChangeMsg
+    hasLedChange -> ledDateChangeMsg
+    hasArdOrCrdChanged -> crdOrArdChangeMsg
+    else -> null
+  }
+}
+private fun isAdditionalConditionOf14B(
+  conditionCode: String
+) = conditionCode == conditionCode14B
+
+fun calculateEndDate(
```

Comment:
rather than talking about condition14b can we talk about electronic monitoring?

`endDate` -> `electronicMonitoringEndDate`
`calculateEndDate` -> `calculateElectronicMonitoringEndDate`
`isElectronicMonitoringConditionPresent`

etc?

Reply:
done